android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothStandardWeightProfile.java

```diff
@@ -213,91 +385,435 @@ private void handleBodyCompositionMeasurement(byte[] value) {
+        return calendar;
+    }
+
+    protected void writeGender() {
+        BluetoothBytesParser parser = new BluetoothBytesParser();
+        int gender = this.selectedUser.getGender().toInt();
+        Timber.d(String.format("gender: %d", gender));
+        parser.setIntValue(gender, FORMAT_UINT8);
+        writeBytes(BluetoothGattUuid.SERVICE_USER_DATA, BluetoothGattUuid.CHARACTERISTIC_USER_GENDER,
+                parser.getValue());
+    }
+
+    protected void writeHeight() {
+        BluetoothBytesParser parser = new BluetoothBytesParser();
+        int height = (int) this.selectedUser.getBodyHeight();
+        Timber.d(String.format("height: %d", height));
+        parser.setIntValue(height, FORMAT_UINT16);
+        writeBytes(BluetoothGattUuid.SERVICE_USER_DATA, BluetoothGattUuid.CHARACTERISTIC_USER_HEIGHT,
+                parser.getValue());
+    }
+
+    protected void writeActivityLevel() {
+        Timber.d("Write user activity level not implemented!");
+    }
+
+    protected void setChangeIncrement() {
+        BluetoothBytesParser parser = new BluetoothBytesParser();
+        parser.setIntValue(1, FORMAT_UINT8);
+        writeBytes(BluetoothGattUuid.SERVICE_USER_DATA, BluetoothGattUuid.CHARACTERISTIC_CHANGE_INCREMENT,
+                parser.getValue());
+    }
+
+    protected void requestMeasurement() {
+        Timber.d("Take measurement command not implemented!");
+    }
+
+    protected synchronized void storeUserScaleConsentCode(int userId, int consentCode) {
+        prefs.edit().putInt("userConsentCode" + userId, consentCode).apply();
+    }
+
+    protected synchronized int getUserScaleConsent(int userId) {
+        return prefs.getInt("userConsentCode" + userId, -1);
+    }
+
+    protected synchronized void storeUserScaleIndex(int userId, int userIndex) {
+        int currentUserIndex = getUserScaleIndex(userId);
+        if (currentUserIndex != -1) {
+            prefs.edit().putInt("userIdFromUserScaleIndex" + currentUserIndex, -1);
+        }
+        prefs.edit().putInt("userScaleIndex" + userId, userIndex).apply();
+        if (userIndex != -1) {
+            prefs.edit().putInt("userIdFromUserScaleIndex" + userIndex, userId).apply();
+        }
+    }
+
+    protected synchronized int getUserIdFromScaleIndex(int userScaleIndex) {
+        return prefs.getInt("userIdFromUserScaleIndex" + userScaleIndex, -1);
+    }
+
+    protected synchronized int getUserScaleIndex(int userId) {
+        return prefs.getInt("userScaleIndex" + userId, -1);
+    }
+
+    protected void reconnectOrSetSmState(SM_STEPS requestedState, SM_STEPS minState, Handler uiHandler) {
+        if (needReConnect()) {
+            jumpNextToStepNr(SM_STEPS.START.ordinal());
+            stopMachineState();
+            reConnectPreviousPeripheral(uiHandler);
+            return;
+        }
+        if (getStepNr() > minState.ordinal()) {
+            jumpNextToStepNr(requestedState.ordinal());
+        }
+        resumeMachineState();
+    }
```

Comment:
Why is this method needed? Could we avoid this method? Is the problem that the scale needs a reconnect?

Reply:
This method is needed to initiate re-connection to Bluetooth device after user made a scale-user-selection or entered scale user consent.
It may take more then a minute for user to figure out how to display consent code on scale OR to choose which scale-user to use.
The sequence is something like this:

1. user clicks Bluetooth icon in OpenScale.
2. State machine in BluetoothGattUuidBF600 class (or other class that extends BluetoothStandardWeightProfile) will read user list from scale.
3. If there is no user index stored in SharedPreferences, then send CHOOSE_SCALE_USER message with list of users to MainActivity;
4. MainActivity will create AlertDialog with list of users (and option to create new user);
5. If user spends more than 60 seconds in this dialo then Bluetooth will disconnect;
6. When user finally makes decision onClick() callback will send selection back to OpenScale object. 
7. OpenScale object in turn knows about exact scale class that has been used and pass that information further to BluetoothGattUuidBF600 object;
8. reconnectOrSetSmState() then checks if Bluetooth is still connected. If it is still connected, then set state-machine state and resume state-machine. If it is not connected then reset state-machine to `start` and initiate re-connection to the same Bluetooth device.